docs/WASI-design-principles.md

```diff
@@ -0,0 +1,95 @@
+# WASI Design Principles
+
+## Capability-based security
+
+WASI is built using capability-based security principles. Access to external resources is always represented by *handles*, which are special values that are *unforgeable*, meaning there's no way to coerce an arbitrary integer or other type of value into a handle. WASI is also aiming to have no *ambient authorities*, meaning that there should be no way to request a handle purely by providing a string or other user-controlled identifier providing the name of a resource. With these two properties, the only ways to obtain access to resources are to be explicitly given handles, or to perform operations on handles which return new handles.
+
+Note that this is a different sense of "capability" than [Linux capabilities](http://man7.org/linux/man-pages/man7/capabilities.7.html) or the withdrawn [POSIX capabilities](https://archive.org/details/posix_1003.1e-990310), which are per-process rather than per-resource.
+
+The simplest representation of handles are values of [reference type](https://github.com/WebAssembly/reference-types). References in wasm are inherently unforgeable, so they can represent handles directly.
+
+Some programming languages operate primarily within linear memory, such as C, C++, and Rust, and there currently is no easy way for these languages to use references in normal code. And even if it does become possible, it's likely that source code will still require annotations to fully opt into references, so it won't always be feasible to use. For these languages, references are stored in a [table](https://webassembly.github.io/spec/core/bikeshed/index.html#tables%E2%91%A0) called a *c-list*. Integer indices into the table then identify resources, which can be easily passed around or stored in memory. In some contexts, these indices are called *file descriptors* since they're similar to what POSIX uses that term for. There are even some tools, such as wasm-bindgen, which make this fairly easy. (Internally, tools and engines don't always use actual WebAssembly tables to do this, however those are implementation details. Conceptually, they work as if they had tables.)
```

Comment:
Again, we should be clear up front that this doesn't actually describe the current implemenation... I see now that you do this below which maybe is fine, but it means that people who just glance at this section might get the wrong impression.

Reply:
I'm not clear on what you mean. The parenthetical "Internally, ..." does describe what effectively happens today -- there is no actual WebAssembly table, but things act as if there were.

Reply:
Agreed.